File: sdk/ruby/lib/deploysentry/streaming.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"

module DeploySentry
  class SSEClient
    INITIAL_RETRY_DELAY = 1
    MAX_RETRY_DELAY = 30
    BACKOFF_MULTIPLIER = 2
    JITTER_FRACTION = 0.2

    def initialize(url:, headers: {}, on_event: nil, on_error: nil)
      @url = URI(url)
      @headers = headers
      @on_event = on_event
      @on_error = on_error
      @running = false
      @thread = nil
      @mutex = Mutex.new
    end

    def start
      @mutex.synchronize do
        return if @running

        @running = true
        @thread = Thread.new { run_loop }
        @thread.abort_on_exception = false
        @thread.name = "deploysentry-sse"
      end
    end

    def stop
      @mutex.synchronize do
        @running = false
      end
      @thread&.join(5)
      @thread = nil
    end

    def running?
      @mutex.synchronize { @running }
    end

    private

    def run_loop
      delay = INITIAL_RETRY_DELAY

      while running?
        begin
          connect_and_stream
          delay = INITIAL_RETRY_DELAY
        rescue => e
          notify_error(e) if running?
          if running?
            jitter = delay * JITTER_FRACTION * (2 * rand - 1)
            jittered = [delay + jitter, 0].max
            sleep(jittered)
            delay = [delay * BACKOFF_MULTIPLIER, MAX_RETRY_DELAY].min
          end
        end
      end
    end

    def connect_and_stream
      http = Net::HTTP.new(@url.host, @url.port)
      http.use_ssl = @url.scheme == "https"
      http.open_timeout = 10
      http.read_timeout = 0

      request = Net::HTTP::Get.new(@url)
      @headers.each { |k, v| request[k] = v }
      request["Accept"] = "text/event-stream"
      request["Cache-Control"] = "no-cache"

      http.request(request) do |response|
        unless response.is_a?(Net::HTTPSuccess)
          raise ConnectionError, "SSE connection failed: #{response.code} #{response.message}"
        end

        buffer = ""

        response.read_body do |chunk|
          break unless running?

          buffer += chunk
          while (idx = buffer.index("\n\n"))
            raw_event = buffer[0...idx]
            buffer = buffer[(idx + 2)..]
            parse_and_dispatch(raw_event)
          end
        end
      end
    ensure
      http&.finish if http&.started?
    end

    def parse_and_dispatch(raw)
      event_type = nil
      data_lines = []

      raw.each_line do |line|
        line = line.chomp
        if line.start_with?("event:")
          event_type = line.sub("event:", "").strip
        elsif line.start_with?("data:")
          data_lines << line.sub("data:", "").strip
        end
      end

      return if data_lines.empty?

      data = data_lines.join("\n")
      parsed = begin
        JSON.parse(data, symbolize_names: true)
      rescue JSON::ParserError
        data
      end

      @on_event&.call(event_type || "message", parsed)
    end

    def notify_error(error)
      @on_error&.call(error)
    end
  end
end

File: sdk/python/deploysentry/streaming.py
"""Server-Sent Events (SSE) streaming client for real-time flag updates."""

from __future__ import annotations

import asyncio
import json
import logging
import random
import threading
import time
from typing import Any, Callable, Dict, Optional

import httpx

logger = logging.getLogger("deploysentry.streaming")

# Reconnection parameters
_INITIAL_RETRY_MS = 1000
_MAX_RETRY_MS = 30000
_RETRY_MULTIPLIER = 2


def _apply_jitter(delay_ms: float) -> float:
    """Apply +/- 20% jitter to a delay value."""
    jitter = delay_ms * 0.2 * (2 * random.random() - 1)
    return delay_ms + jitter


class SSEClient:
    """Synchronous SSE client that runs in a background thread.

    Connects to the DeploySentry flag-stream endpoint, parses SSE frames, and
    invokes *on_update* for every ``data:`` message received.
    """

    def __init__(
        self,
        url: str,
        headers: Dict[str, str],
        on_update: Callable[[Dict[str, Any]], None],
        params: Optional[Dict[str, str]] = None,
    ) -> None:
        self._url = url
        self._headers = headers
        self._on_update = on_update
        self._params = params or {}
        self._stop_event = threading.Event()
        self._thread: Optional[threading.Thread] = None

    def start(self) -> None:
        """Start the SSE listener in a daemon thread."""
        if self._thread is not None and self._thread.is_alive():
            return
        self._stop_event.clear()
        self._thread = threading.Thread(target=self._run, daemon=True, name="deploysentry-sse")
        self._thread.start()

    def stop(self) -> None:
        """Signal the listener to stop and wait for thread termination."""
        self._stop_event.set()
        if self._thread is not None:
            self._thread.join(timeout=5)
            self._thread = None

    @property
    def is_running(self) -> bool:
        return self._thread is not None and self._thread.is_alive()

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _run(self) -> None:
        retry_ms = _INITIAL_RETRY_MS
        while not self._stop_event.is_set():
            try:
                self._connect_and_listen()
                retry_ms = _INITIAL_RETRY_MS  # reset on clean disconnect
            except Exception:
                jittered = _apply_jitter(retry_ms)
                logger.warning(
                    "SSE connection lost, retrying in %.0f ms", jittered, exc_info=True
                )
                if self._stop_event.wait(timeout=jittered / 1000):
                    break
                retry_ms = min(retry_ms * _RETRY_MULTIPLIER, _MAX_RETRY_MS)

    def _connect_and_listen(self) -> None:
        sse_headers = {**self._headers, "Accept": "text/event-stream"}
        with httpx.Client(timeout=None) as http:
            with http.stream("GET", self._url, headers=sse_headers, params=self._params) as resp:
                resp.raise_for_status()
                buffer = ""
                for chunk in resp.iter_text():
                    if self._stop_event.is_set():
                        return
                    buffer += chunk
                    while "\n\n" in buffer:
                        frame, buffer = buffer.split("\n\n", 1)
                        self._process_frame(frame)

    def _process_frame(self, frame: str) -> None:
        event_type = ""
        data_lines: list[str] = []
        for line in frame.split("\n"):
            if line.startswith("event:"):
                event_type = line[len("event:"):].strip()
            elif line.startswith("data:"):
                data_lines.append(line[len("data:"):].strip())
            elif line.startswith("retry:"):
                # Server-sent retry hint; ignored for simplicity.
                pass
        if not data_lines:
            return
        raw = "\n".join(data_lines)
        try:
            payload = json.loads(raw)
        except json.JSONDecodeError:
            logger.debug("Non-JSON SSE data: %s", raw)
            return
        self._on_update(payload)


class AsyncSSEClient:
    """Asynchronous SSE client backed by ``asyncio``."""

    def __init__(
        self,
        url: str,
        headers: Dict[str, str],
        on_update: Callable[[Dict[str, Any]], None],
        params: Optional[Dict[str, str]] = None,
    ) -> None:
        self._url = url
        self._headers = headers
        self._on_update = on_update
        self._params = params or {}
        self._task: Optional[asyncio.Task[None]] = None

    async def start(self) -> None:
        """Start the SSE listener as an asyncio task."""
        if self._task is not None and not self._task.done():
            return
        self._task = asyncio.create_task(self._run())

    async def stop(self) -> None:
        """Cancel the listener task and await its completion."""
        if self._task is not None:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
            self._task = None

    @property
    def is_running(self) -> bool:
        return self._task is not None and not self._task.done()

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    async def _run(self) -> None:
        retry_ms = _INITIAL_RETRY_MS
        while True:
            try:
                await self._connect_and_listen()
                retry_ms = _INITIAL_RETRY_MS
            except asyncio.CancelledError:
                raise
            except Exception:
                jittered = _apply_jitter(retry_ms)
                logger.warning(
                    "SSE connection lost, retrying in %.0f ms", jittered, exc_info=True
                )
                await asyncio.sleep(jittered / 1000)
                retry_ms = min(retry_ms * _RETRY_MULTIPLIER, _MAX_RETRY_MS)

    async def _connect_and_listen(self) -> None:
        sse_headers = {**self._headers, "Accept": "text/event-stream"}
        async with httpx.AsyncClient(timeout=None) as http:
            async with http.stream(
                "GET", self._url, headers=sse_headers, params=self._params
            ) as resp:
                resp.raise_for_status()
                buffer = ""
                async for chunk in resp.aiter_text():
                    buffer += chunk
                    while "\n\n" in buffer:
                        frame, buffer = buffer.split("\n\n", 1)
                        self._process_frame(frame)

    def _process_frame(self, frame: str) -> None:
        event_type = ""
        data_lines: list[str] = []
        for line in frame.split("\n"):
            if line.startswith("event:"):
                event_type = line[len("event:"):].strip()
            elif line.startswith("data:"):
                data_lines.append(line[len("data:"):].strip())
        if not data_lines:
            return
        raw = "\n".join(data_lines)
        try:
            payload = json.loads(raw)
        except json.JSONDecodeError:
            logger.debug("Non-JSON SSE data: %s", raw)
            return
        self._on_update(payload)
